File: frontend/src/lib/tenant-readiness.ts
import { ApiClientError } from "./api-client";
import { requestAuthenticatedApi } from "./session";

export const TENANT_READINESS_ITEM_KEYS = [
  "default_legal_entity",
  "organization_structure",
  "active_tenant_administrator",
  "employee_master_data",
  "leave_configuration",
  "document_configuration",
  "privacy_notice",
  "feature_dependencies",
  "notification_delivery",
] as const;

export const TENANT_READINESS_ITEM_STATES = [
  "ready",
  "action_required",
  "not_applicable",
] as const;

export type TenantReadinessItemKey =
  (typeof TENANT_READINESS_ITEM_KEYS)[number];
export type TenantReadinessItemState =
  (typeof TENANT_READINESS_ITEM_STATES)[number];
export type TenantReadinessOverallState = "ready" | "action_required";
export type TenantReadinessRemediationRoute =
  | "/organization"
  | "/users"
  | "/employees"
  | "/leave/admin"
  | "/document-types"
  | "/privacy/manage";

export interface TenantReadinessItem {
  key: TenantReadinessItemKey;
  state: TenantReadinessItemState;
  count: number | null;
  remediation_route: TenantReadinessRemediationRoute | null;
  evaluated_at: string;
}

export interface TenantReadiness {
  overall_state: TenantReadinessOverallState;
  evaluated_at: string;
  items: TenantReadinessItem[];
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function hasExactKeys(
  value: Record<string, unknown>,
  expected: readonly string[],
): boolean {
  const actual = Object.keys(value);
  return (
    actual.length === expected.length &&
    expected.every((key) => Object.hasOwn(value, key))
  );
}

function isUtcDateTime(value: unknown): value is string {
  return (
    typeof value === "string" &&
    /(?:Z|[+-]00:00)$/.test(value) &&
    Number.isFinite(Date.parse(value))
  );
}

function isItemState(value: unknown): value is TenantReadinessItemState {
  return (
    typeof value === "string" &&
    TENANT_READINESS_ITEM_STATES.includes(
      value as TenantReadinessItemState,
    )
  );
}

function isOverallState(value: unknown): value is TenantReadinessOverallState {
  return value === "ready" || value === "action_required";
}

function isCountForItem(
  key: TenantReadinessItemKey,
  value: unknown,
): value is number | null {
  if (value === null) {
    return (
      key === "organization_structure" ||
      key === "leave_configuration" ||
      key === "feature_dependencies" ||
      key === "notification_delivery"
    );
  }
  if (
    typeof value !== "number" ||
    !Number.isSafeInteger(value) ||
    value < 0
  ) {
    return false;
  }

  switch (key) {
    case "default_legal_entity":
    case "privacy_notice":
      return value <= 1;
    case "active_tenant_administrator":
    case "employee_master_data":
    case "document_configuration":
      return true;
    case "organization_structure":
    case "leave_configuration":
    case "feature_dependencies":
    case "notification_delivery":
      return false;
  }
}

function isStateAllowedForItem(
  key: TenantReadinessItemKey,
  state: TenantReadinessItemState,
): boolean {
  if (key === "notification_delivery" && state === "ready") {
    return false;
  }
  if (state !== "not_applicable") {
    return true;
  }
  return (
    key === "leave_configuration" ||
    key === "document_configuration" ||
    key === "notification_delivery"
  );
}

function isRouteForItem(
  key: TenantReadinessItemKey,
  value: unknown,
  count: number | null,
): value is TenantReadinessRemediationRoute | null {
  switch (key) {
    case "default_legal_entity":
    case "organization_structure":
      return value === "/organization";
    case "active_tenant_administrator":
      return value === "/users";
    case "employee_master_data":
      return value === "/employees";
    case "leave_configuration":
      return value === "/leave/admin";
    case "document_configuration":
      return count === 0
        ? value === "/document-types"
        : value === null;
    case "privacy_notice":
      return value === "/privacy/manage";
    case "feature_dependencies":
    case "notification_delivery":
      return value === null;
  }
}

function isStateConsistentWithCount(
  key: TenantReadinessItemKey,
  state: TenantReadinessItemState,
  count: number | null,
): boolean {
  switch (key) {
    case "default_legal_entity":
      return (count === 1) === (state === "ready");
    case "active_tenant_administrator":
    case "employee_master_data":
    case "privacy_notice":
      return ((count ?? 0) > 0) === (state === "ready");
    case "document_configuration":
      return count !== 0 || state !== "ready";
    case "organization_structure":
    case "leave_configuration":
    case "feature_dependencies":
    case "notification_delivery":
      return true;
  }
}

function isReadinessItem(
  value: unknown,
  expectedKey: TenantReadinessItemKey,
  evaluatedAt: string,
): value is TenantReadinessItem {
  if (
    !isRecord(value) ||
    !hasExactKeys(value, [
      "key",
      "state",
      "count",
      "remediation_route",
      "evaluated_at",
    ]) ||
    value.key !== expectedKey ||
    !isItemState(value.state) ||
    !isStateAllowedForItem(expectedKey, value.state) ||
    !isCountForItem(expectedKey, value.count) ||
    !isRouteForItem(expectedKey, value.remediation_route, value.count) ||
    !isStateConsistentWithCount(expectedKey, value.state, value.count) ||
    !isUtcDateTime(value.evaluated_at)
  ) {
    return false;
  }
  return value.evaluated_at === evaluatedAt;
}

function isTenantReadiness(value: unknown): value is TenantReadiness {
  if (
    !isRecord(value) ||
    !hasExactKeys(value, ["overall_state", "evaluated_at", "items"]) ||
    !isOverallState(value.overall_state) ||
    !isUtcDateTime(value.evaluated_at) ||
    !Array.isArray(value.items) ||
    value.items.length !== TENANT_READINESS_ITEM_KEYS.length
  ) {
    return false;
  }

  for (let index = 0; index < TENANT_READINESS_ITEM_KEYS.length; index += 1) {
    if (
      !isReadinessItem(
        value.items[index],
        TENANT_READINESS_ITEM_KEYS[index],
        value.evaluated_at,
      )
    ) {
      return false;
    }
  }

  const items = value.items as TenantReadinessItem[];
  const expectedOverallState: TenantReadinessOverallState = items.every(
    (item) => item.state === "ready" || item.state === "not_applicable",
  )
    ? "ready"
    : "action_required";
  return value.overall_state === expectedOverallState;
}

export async function readTenantReadiness(): Promise<TenantReadiness> {
  const data = await requestAuthenticatedApi<unknown>(
    "/api/v1/tenant/readiness",
  );
  if (!isTenantReadiness(data)) {
    throw new ApiClientError({ status: 200, code: "invalid_response" });
  }
  return data;
}

File: backend/app/schemas/tenant_readiness.py
"""Strict tenant setup-readiness response contract."""

from __future__ import annotations

from datetime import timedelta
from enum import StrEnum
from typing import Annotated, Literal, Self

from pydantic import (
    AwareDatetime,
    BaseModel,
    ConfigDict,
    Field,
    field_validator,
    model_validator,
)


class TenantReadinessItemKey(StrEnum):
    DEFAULT_LEGAL_ENTITY = "default_legal_entity"
    ORGANIZATION_STRUCTURE = "organization_structure"
    ACTIVE_TENANT_ADMINISTRATOR = "active_tenant_administrator"
    EMPLOYEE_MASTER_DATA = "employee_master_data"
    LEAVE_CONFIGURATION = "leave_configuration"
    DOCUMENT_CONFIGURATION = "document_configuration"
    PRIVACY_NOTICE = "privacy_notice"
    FEATURE_DEPENDENCIES = "feature_dependencies"
    NOTIFICATION_DELIVERY = "notification_delivery"


class TenantReadinessItemState(StrEnum):
    READY = "ready"
    ACTION_REQUIRED = "action_required"
    NOT_APPLICABLE = "not_applicable"


class TenantReadinessOverallState(StrEnum):
    READY = "ready"
    ACTION_REQUIRED = "action_required"


TenantReadinessCount = Annotated[int, Field(strict=True, ge=0)]
TenantReadinessRemediationRoute = Literal[
    "/organization",
    "/users",
    "/employees",
    "/leave/admin",
    "/document-types",
    "/privacy/manage",
]

TENANT_READINESS_ITEM_ORDER: tuple[TenantReadinessItemKey, ...] = tuple(TenantReadinessItemKey)

_COUNTED_ITEM_KEYS = frozenset(
    {
        TenantReadinessItemKey.DEFAULT_LEGAL_ENTITY,
        TenantReadinessItemKey.ACTIVE_TENANT_ADMINISTRATOR,
        TenantReadinessItemKey.EMPLOYEE_MASTER_DATA,
        TenantReadinessItemKey.DOCUMENT_CONFIGURATION,
        TenantReadinessItemKey.PRIVACY_NOTICE,
    }
)
_OPTIONAL_ITEM_KEYS = frozenset(
    {
        TenantReadinessItemKey.LEAVE_CONFIGURATION,
        TenantReadinessItemKey.DOCUMENT_CONFIGURATION,
        TenantReadinessItemKey.NOTIFICATION_DELIVERY,
    }
)
_FIXED_REMEDIATION_ROUTES: dict[
    TenantReadinessItemKey,
    TenantReadinessRemediationRoute | None,
] = {
    TenantReadinessItemKey.DEFAULT_LEGAL_ENTITY: "/organization",
    TenantReadinessItemKey.ORGANIZATION_STRUCTURE: "/organization",
    TenantReadinessItemKey.ACTIVE_TENANT_ADMINISTRATOR: "/users",
    TenantReadinessItemKey.EMPLOYEE_MASTER_DATA: "/employees",
    TenantReadinessItemKey.LEAVE_CONFIGURATION: "/leave/admin",
    TenantReadinessItemKey.PRIVACY_NOTICE: "/privacy/manage",
    TenantReadinessItemKey.FEATURE_DEPENDENCIES: None,
    TenantReadinessItemKey.NOTIFICATION_DELIVERY: None,
}


class TenantReadinessItemRead(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    key: TenantReadinessItemKey
    state: TenantReadinessItemState
    count: TenantReadinessCount | None
    remediation_route: TenantReadinessRemediationRoute | None
    evaluated_at: AwareDatetime

    @field_validator("evaluated_at")
    @classmethod
    def require_utc_timestamp(cls, value: AwareDatetime) -> AwareDatetime:
        if value.utcoffset() != timedelta(0):
            raise ValueError("Readiness timestamps must be UTC")
        return value

    @model_validator(mode="after")
    def require_item_contract(self) -> Self:
        if (self.key in _COUNTED_ITEM_KEYS) != (self.count is not None):
            raise ValueError("Readiness count presence does not match the item contract")
        if self.key in {
            TenantReadinessItemKey.DEFAULT_LEGAL_ENTITY,
            TenantReadinessItemKey.PRIVACY_NOTICE,
        } and self.count not in {0, 1}:
            raise ValueError("This readiness count must be zero or one")
        if (
            self.state is TenantReadinessItemState.NOT_APPLICABLE
            and self.key not in _OPTIONAL_ITEM_KEYS
        ):
            raise ValueError("This readiness item is never not-applicable")
        if (
            self.key is TenantReadinessItemKey.NOTIFICATION_DELIVERY
            and self.state is TenantReadinessItemState.READY
        ):
            raise ValueError("Notification delivery cannot be pilot-ready")

        expected_route = (
            "/document-types"
            if self.key is TenantReadinessItemKey.DOCUMENT_CONFIGURATION and self.count == 0
            else None
            if self.key is TenantReadinessItemKey.DOCUMENT_CONFIGURATION
            else _FIXED_REMEDIATION_ROUTES[self.key]
        )
        if self.remediation_route != expected_route:
            raise ValueError("Readiness remediation route does not match the item contract")

        if self.key is TenantReadinessItemKey.DEFAULT_LEGAL_ENTITY and (
            (self.count == 1) != (self.state is TenantReadinessItemState.READY)
        ):
            raise ValueError("Default legal-entity readiness does not match its count")
        if self.key in {
            TenantReadinessItemKey.ACTIVE_TENANT_ADMINISTRATOR,
            TenantReadinessItemKey.EMPLOYEE_MASTER_DATA,
            TenantReadinessItemKey.PRIVACY_NOTICE,
        } and ((self.count or 0) > 0) != (self.state is TenantReadinessItemState.READY):
            raise ValueError("Readiness state does not match its resource count")
        if (
            self.key is TenantReadinessItemKey.DOCUMENT_CONFIGURATION
            and self.count == 0
            and self.state is TenantReadinessItemState.READY
        ):
            raise ValueError("Document readiness requires a document type")
        return self


class TenantReadinessRead(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    overall_state: TenantReadinessOverallState
    evaluated_at: AwareDatetime
    items: list[TenantReadinessItemRead] = Field(
        min_length=len(TENANT_READINESS_ITEM_ORDER),
        max_length=len(TENANT_READINESS_ITEM_ORDER),
    )

    @field_validator("evaluated_at")
    @classmethod
    def require_utc_timestamp(cls, value: AwareDatetime) -> AwareDatetime:
        if value.utcoffset() != timedelta(0):
            raise ValueError("Readiness timestamps must be UTC")
        return value

    @model_validator(mode="after")
    def require_projection_contract(self) -> Self:
        if tuple(item.key for item in self.items) != TENANT_READINESS_ITEM_ORDER:
            raise ValueError("Readiness items must use the fixed checklist order")
        if any(item.evaluated_at != self.evaluated_at for item in self.items):
            raise ValueError("All readiness timestamps must be identical")
        expected_overall_state = (
            TenantReadinessOverallState.READY
            if all(
                item.state
                in {
                    TenantReadinessItemState.READY,
                    TenantReadinessItemState.NOT_APPLICABLE,
                }
                for item in self.items
            )
            else TenantReadinessOverallState.ACTION_REQUIRED
        )
        if self.overall_state is not expected_overall_state:
            raise ValueError("Overall readiness does not match the checklist")
        return self


__all__ = [
    "TENANT_READINESS_ITEM_ORDER",
    "TenantReadinessItemKey",
    "TenantReadinessItemRead",
    "TenantReadinessItemState",
    "TenantReadinessOverallState",
    "TenantReadinessRead",
]
